### ISSUE TITLE
ConfigProperties no prefix bug

### ISSUE THREAD

**Issue by Joseph-Cass** [smallrye/smallrye-config#475]
Injecting a `ConfigProperties` class, [without a prefix defined](https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#_inject_the_bean_annotated_with_code_configproperties_code), leads to `io.smallrye.config.inject.ConfigExtension.validate` checking for the desired Config Property without the intended prefix included. If this incorrect Config Property check isn't found a `java.util.NoSuchElementException` is thrown.

For example, Injecting:
```
@Inject
@ConfigProperties
ConfigPropertiesPOJO details;
```

Where ConfigPropertiesPOJO:
```
@ConfigProperties(prefix = "myPrefix")
@Dependent
public class ConfigPropertiesPOJO {
    public String validkey;
}
```

With the following in a Config Source:
```
myPrefix.validkey=validValue
```

Leads to checks for `myPrefix.validkey` (intentionally) AND `validkey` (unintentionally) which throws `java.util.NoSuchElementException: SRCFG00014: Property validkey not found`.

This isn't thrown in the TCKs since in all the examples both `<prefix>.<key>` and `<key>` are unintentionally defined in the Config Sources. E.g. in `customer.name ` and `name ` in [ConfigPropertiesTest.java](https://github.com/eclipse/microprofile-config/blob/93943308fefcca5d5b484ff3b1f8e080589f64ab/tck/src/main/java/org/eclipse/microprofile/config/tck/ConfigPropertiesTest.java#L100).

This issue did not occur in a previous iteration of smallrye-config:2.0.0 (though I can't remember which one... perhaps as far back as RC1?)

**Comment by radcortez** [comment]
There was some back and forward regarding the prefix override. If I remember correctly only if you explicitly add an empty string in the injection point overrides the prefix to no prefix. Without a prefix in the annotating it should fallback to the prefix set in the properties class. 

I'll have a look.

**Comment by Emily-Jiang** [comment]
@radcortez the absence of prefix on the injection point of @ConfigProperties means the prefix defined on the ConfigProperties bean used. In this issue raised by @Joseph-Cass , the lookup should use myPrefix as the prefix.

**Comment by Emily-Jiang** [comment]
Basically it should return after myPrefix.validKey was found. It looks like the recursive check is redundant and causes this issue.

**Comment by radcortez** [comment]
I was not able to reproduce this. We have a test for it. I've done some refactoring there to split it from other tests: https://github.com/smallrye/smallrye-config/pull/481. Can you please check?

The validation only kicks in when an instance of the object is created and for that to happen and injection point with an empty prefix must be present in the CDI container: https://github.com/smallrye/smallrye-config/blob/e575006dd92dc7ebbcc4cc742716c04e655d55be/cdi/src/main/java/io/smallrye/config/inject/ConfigExtension.java#L177.

Maybe I'm missing something. Are you able to provide me with a reproducer?

**Comment by Emily-Jiang** [comment]
@radcortez IIUC, In order to recreate the issue, in your refactored test https://github.com/smallrye/smallrye-config/blob/master/cdi/src/test/java/io/smallrye/config/inject/ConfigPropertiesInjectionTest.java#L46, just set `server.theHost` and `cloud.theHost` in the config source without adding `theHost` property. I was trying to reproduce this but I could not locate where the properties were set. Can you point to the place where the properties was added?

**Comment by Joseph-Cass** [comment]
> @radcortez IIUC, In order to recreate the issue, in your refactored test https://github.com/smallrye/smallrye-config/blob/master/cdi/src/test/java/io/smallrye/config/inject/ConfigPropertiesInjectionTest.java#L46, just set `server.theHost` and `cloud.theHost` in the config source without adding `theHost` property. I was trying to reproduce this but I could not locate where the properties were set. Can you point to the place where the properties was added?

I think `theHost` is not added. I added `assertThrows(NoSuchElementException.class, () -> config.getValue("theHost", String.class));` to https://github.com/smallrye/smallrye-config/blob/aed59f138e2c876947a927e5f35b36a88280a734/cdi/src/test/java/io/smallrye/config/inject/ConfigPropertiesInjectionTest.java#L44
and the test still passed. 

Perhaps this issue was fixed between versions. I will re-consume the latest SmallRye version tomorrow, and test if the expected bahaviour occurs.

Thanks!

**Comment by radcortez** [comment]
Yes, I'm also asserting that these values are not set here:
https://github.com/smallrye/smallrye-config/blob/3df12dd156205fb988fa05246b5813356d1b385a/cdi/src/test/java/io/smallrye/config/inject/ConfigPropertiesInjectionTest.java#L69-L70

The configuration values are set here:
https://github.com/smallrye/smallrye-config/blob/3df12dd156205fb988fa05246b5813356d1b385a/cdi/src/test/java/io/smallrye/config/inject/InjectionTestConfigFactory.java#L21-L56

**Comment by radcortez** [comment]
> Perhaps this issue was fixed between versions. I will re-consume the latest SmallRye version tomorrow, and test if the expected bahaviour occurs.

Ok, let me know the result and we try to figure out if something is wrong.

**Comment by Emily-Jiang** [comment]
Thanks @radcortez for the explaination! Should you check 
`assertNull(config.getConfigValue("theHost").getValue()); `
?

**Comment by radcortez** [comment]
Added.

**Comment by Joseph-Cass** [comment]
So I think I figured out what was going wrong. This test case was failing against a locally built (very) old version of SmallRye Config from as far back as October, a version preceding 2.0.0-RC1. It must have been fixed before the release of RC1 as I couldn't replicate it against anything from 2.0.0-RC1 or later.

https://github.com/OpenLiberty/open-liberty/pull/15494 confirms that the 3 ConfigProperties issues raised are no longer issues against the most recent version of SmallRye Config.

Thanks for your help!
